src/oc/storage/async/notification.clj

```diff
@@ -0,0 +1,142 @@
+
+(defn resource-type? [resource-type]
+  (or (= resource-type :org) (= resource-type :board) (= resource-type :entry)))
+
+(def NotificationTrigger
+  "
+  A trigger for one of the various types of notifications that are published:
+
+  add - the content in  newly created, this happens when a board or entry is added
+  update - the content-id should be refreshed, this happens when a board or entry is updated
+  delete - the specified content-id is deleted, this happens when a board or entry is removed
+
+  The notification trigger contains the type of resource as `resource-type` and the current resource as `current`. If
+  the trigger is for an update it will also contain the updated resource as `update`. 
+
+  The user whose actions triggered the notification is included as `user`.
+
+  A timestamp for when the notice was created is included as `notice-at`.
+  "
+  {:notification-type (schema/pred notification-type?)
+   :resource-type (schema/pred resource-type?)
+   :current (schema/conditional #(= (resource-type %) :entry) common-res/Entry
+                                #(= (resource-type %) :board) common-res/Board
+                                :else common-res/Org)
+   (schema/optional-key :update) (schema/conditional #(= (resource-type %) :entry) common-res/Entry
+                                                     #(= (resource-type %) :board) common-res/Board
+                                                     :else common-res/Org)
+   :user lib-schema/User
+   :notification-at lib-schema/ISO8601})
+
+;; ----- Event handling -----
+
+(defn handle-notification-message
+  [trigger]
+  (timbre/debug "Notification request of:" (:notification-type trigger)
+               "for:" (trigger :current :uuid) "to topic:" config/aws-sns-storage-topic-arn)
+  (timbre/trace "Notification request:" trigger)
+  (schema/validate NotificationTrigger trigger)
+  (timbre/info "Sending request to topic:" config/aws-sns-storage-topic-arn)
+  (sns/publish
+    {:access-key config/aws-access-key-id
+     :secret-key config/aws-secret-access-key}
+     :topic-arn config/aws-sns-storage-topic-arn
+     :subject (str (name (:notification-type trigger))
+                   " on " (name (:resource-type trigger))
+                   ": " (-> trigger :current :uuid))
+     :message (json/generate-string trigger {:pretty true}))
+  (timbre/info "Request sent to topic:" config/aws-sns-storage-topic-arn))
+
+;; ----- Event loop -----
+
+(defn notification-loop []
+  (reset! notification-go true)
+  (timbre/info "Starting notification...")
+  (async/go (while @notification-go
+    (timbre/debug "Notification waiting...")
+    (let [message (<!! notification-chan)]
+      (timbre/debug "Processing message on notification channel...")
+      (if (:stop message)
+        (do (reset! notification-go false) (timbre/info "Notification stopped."))
+        (async/thread
+          (try
+            (handle-notification-message message)
+          (catch Exception e
```

Comment:
Does this cause an infinite loop since `notification-go` is never set to `false`?

Reply:
Yes, the intent is an infinite loop that blocks on receiving a message through the core.async channel (line 90). `notification-go` is set to false when the loop receives a `:stop` message, thus stopping the channel consumer (line 92-93).

You can try this at the REPL with `(go)` and `(stop)`.

Reply:
Ok, my concern is that an `Exception` will be caught and the same message will be processed again.